src/sources/audio.ts

```diff
@@ -53,152 +71,189 @@ export function getRawAudioFingerprint(): number | (() => Promise<number>) {
+        if (context.state === 'suspended') {
+          // The audio context can reject starting until the tab is in foreground. Long fingerprint duration
+          // in background isn't a problem, therefore the retry attempts don't count in background. It can lead to
+          // a situation when a fingerprint takes very long time and finishes successfully. FYI, the audio context
+          // can be suspended when `document.hidden === false` and start running after a retry.
+          if (!document.hidden) {
+            attemptsLeft--
+          }
+          if (attemptsLeft > 0) {
+            setTimeout(tryRender, retryDelay)
+          } else {
+            resolve(null)
+          }
         }
       } catch (error) {
         reject(error)
       }
     }
 
     tryRender()
+  })
+}
 
-    finalize = () => {
-      if (!isFinalized) {
-        isFinalized = true
-        if (startedRunningAt > 0) {
-          startRunningTimeout()
-        }
-      }
+function extractFingerprint(baseSignal: AudioBuffer, clonedSample: Float32Array): number {
+  let fingerprint = undefined
+  let needsDenoising = false
+
+  for (let i = 0; i < clonedSample.length; i += Math.floor(clonedSample.length / 10)) {
+    if (clonedSample[i] === 0) {
+      // In some cases the signal is 0 on a short range for some reason. Ignoring such samples.
+    } else if (fingerprint === undefined) {
+      fingerprint = clonedSample[i]
+    } else if (fingerprint !== clonedSample[i]) {
+      needsDenoising = true
+      break
     }
-  })
+  }
 
-  return [resultPromise, finalize] as const
+  // The looped buffer source works incorrectly in old Safari versions (>14 desktop, >15 mobile).
+  // The looped signal contains only 0s. To fix it, the loop start should be `baseSignal.length - 1.00000000001` and
+  // the loop end should be `baseSignal.length + 0.00000000001` (there can be 10 or 11 0s after the point). But this
+  // solution breaks the looped signal in other browsers. Instead of checking the browser version, we check that the
+  // looped signals comprises only 0s, and if it does, we return the last value of the base signal, because old Safari
+  // versions don't add noise that we want to cancel.
+  if (fingerprint === undefined) {
+    fingerprint = baseSignal.getChannelData(0)[baseSignal.length - 1]
+  } else if (needsDenoising) {
+    fingerprint = getMiddle(clonedSample)
+  }
+
+  return fingerprint
 }
 
-function getHash(signal: ArrayLike<number>): number {
-  let hash = 0
-  for (let i = 0; i < signal.length; ++i) {
-    hash += Math.abs(signal[i])
+/**
+ * Calculates the middle between the minimum and the maximum array item
+ */
+export function getMiddle(signal: ArrayLike<number>): number {
+  let min = Infinity
+  let max = -Infinity
+
+  for (let i = 0; i < signal.length; i++) {
+    const value = signal[i]
+    // In very rare cases the signal is 0 on a short range for some reason. Ignoring such samples.
+    if (value === 0) {
```

Comment:
nit: might be not safe if all the values are zeros (not sure if it's possible) because we will return NaN as a signal value

Reply:
It is indeed possible and handled by `extractFingerprint` (see the comments). `getMiddle` is called only if there is at least 1 non-zero sample in the signal.

Reply:
Ok, makes sense, thanks